doc/dev/cephfs-snapshots.rst

```diff
@@ -53,6 +53,17 @@ To create a CephFS snapshot, create a subdirectory under
 ``.snap`` with a name of your choice. For example, to create a snapshot on
 directory "/1/2/3/", invoke ``mkdir /1/2/3/.snap/my-snapshot-name`` .
 
+.. note::
+   Snapshot names can not start with an underscore ('_'), as these names are
+   reserved for internal usage.
+
```

Comment:
Ok, so I guess the client will need to prepend these with a different character?

Reply:
I'm not sure I understand your question.  These names that start with an underscore are generated by the MDS itself (although I **really** have no idea about how snapshots work and how these special inodes are used internally).

If you're talking about the encrypted snapshot names, which will be base64-encoded strings that may start with the '_' character, prepending another character is probably the best solution.  But TBH I didn't thought a lot about it, as I've been having fun playing whack-a-mole with the MDS snapshots name handling :-)

Reply:
> Then #45208 makes no sense any more, and I will close it.

Well, what I wanted with this note was simply to document the current behaviour, i.e. that a user can not do a `mkdir .snap/_my-snapshot`.  The MDS assumes that snapshots starting with an '_' are "special".

PR #45208 isn't changing this behaviour, it's simply adding an extra flag to make things easier for clients to know an inode is a "long name snapshot" (although, I think the client could figure it out anyway comparing the "<INODE-NUMBER>" part of the name with the actual inode->i_ino).  But I may be missing something...?

Reply:
Currently when creating the snapshots if the snapshot name's first char is '_' it will fail anyway even in fscrypt case with the base64 encode names. So in clients side we can figure out whether the snapshot name is a long snap name or not by checking the name[0].

The next step here is that in kclient side how to properly handle the base64 encoded snapshot names, which may start the '_'. When users just create one human readable snapshot whose name without starting the '_', but it still may fail.

Reply:
> The next step here is that in kclient side how to properly handle the base64 encoded snapshot names, which may start the '_'. When users just create one human readable snapshot whose name without starting the '_', but it still may fail.

Right, that's what I was going to look at next.  But first I'm still trying to make the client handle correctly these long snapshot names when a directory is encrypted.  I've already an experimental patch, but it still needs some cleanup (and more testing).  Hopefully I'll be able to send this patch later today or tomorrow.

Reply:
Yeah, my comment was more about fscrypt. If we have an fscrypted snapshot name, then we'll probably have to prepend it with a legal character.